File: src/day 1/solutions.ts
import { showResult } from "../utils";

const defaultInput = await Bun.file("input.txt").text();

const mapInput = (input: string) => {
	const lSide: number[] = [];
	const rSide: number[] = [];
	input.split("\n").forEach((line) => {
		const mappedLine = line.split(" ").map((d) => Number.parseInt(d));
		// Move the first number to lSide and the last number to rSide
		lSide.push(mappedLine.shift() as number);
		rSide.push(mappedLine.pop() as number);
	});
	return { lSide, rSide };
};

export const findDistance = (input: string = defaultInput) => {
	const { lSide, rSide } = mapInput(input);
	lSide.sort((a, b) => a - b);
	rSide.sort((a, b) => a - b);
	const distance = lSide.reduce(
		(acc, val, i) => acc + Math.abs(rSide[i] - val),
		0,
	);
	return distance;
};

export const findSimilarity = (input: string = defaultInput): number => {
	const { lSide, rSide } = mapInput(input);
	const similarityMap = new Map<number, number>();
	rSide.forEach((val) => {
		similarityMap.set(val, (similarityMap.get(val) ?? 0) + 1);
	});
	const distance = lSide.reduce(
		(acc, val) => acc + val * (similarityMap.get(val) ?? 0),
		0,
	);
	return distance;
};

showResult(findDistance(), 1);
showResult(findSimilarity(), 2);

File: src/day 1/solutions.py
from pyaoc import Env
from collections import Counter

e = Env(1)
e.T("""3   4
4   3
2   5
1   3
3   9
3   3""", 11, 31)


def map_input(input) -> tuple[list[int],list[int]]:
    lines = input.get_valid_lines()
    numbers = [[int(v) for v in line.split()] for line in lines]
    l_side = [x[0] for x in numbers]
    r_side = [x[1] for x in numbers]
    return l_side, r_side


def part1(input):
    l_side, r_side = map_input(input)
    l_side.sort()
    r_side.sort()
    return sum([abs(d[0] - d[1]) for d in zip(l_side, r_side)])


e.run_tests(1, part1)
e.run_main(1, part1)


def part2(input):
    l_side, r_side = map_input(input)
    c = Counter(r_side)
    return sum([v * c[v] for v in l_side if v in c])


e.run_tests(2, part2)
e.run_main(2, part2)